lib/rate_limiting.rb

```diff
@@ -0,0 +1,47 @@
+module RateLimiting
+
+  LOGIN_ATTEMPTS_PERIOD = 1.hour
+  MAX_LOGIN_ATTEMPTS_PER_USER = 12
+  MAX_LOGIN_ATTEMPTS_PER_IP = 10000
+
+  def too_many_log_in_attempts_by_ip?(attempt_period: LOGIN_ATTEMPTS_PERIOD,
+                                      max_attempts: MAX_LOGIN_ATTEMPTS_PER_IP,
+                                      ip:)
+    attempts_considered_since = Time.now - attempt_period
+
+    sign_in_failed_attempts =
+      SecurityLog.sign_in_failed
+                 .where{created_at > attempts_considered_since}
+                 .where(remote_ip: ip)
+                 .count
+
+    login_not_found_attempts =
+      SecurityLog.login_not_found
+                 .where{created_at > attempts_considered_since}
+                 .where(remote_ip: ip)
+                 .count
+
+    sign_in_failed_attempts + login_not_found_attempts >= max_attempts
+  end
+
+  def too_many_log_in_attempts_by_user?(attempt_period: LOGIN_ATTEMPTS_PERIOD,
+                                        max_attempts: MAX_LOGIN_ATTEMPTS_PER_USER,
+                                        user:)
+    oldest_possible_attempts_considered_since = Time.now - attempt_period
+
+    return false if user.nil?
+
+    last_login_time = SecurityLog.sign_in_successful.where(user: user).maximum(:created_at)
+    attempts_considered_since = last_login_time.nil? ?
+                                  oldest_possible_attempts_considered_since :
+                                  [oldest_possible_attempts_considered_since, last_login_time].max
+
+    user_attempts = SecurityLog.sign_in_failed
+                               .where{created_at > attempts_considered_since}
+                               .where(user: user)
+                               .count
+
+    user_attempts >= max_attempts
+  end
+
+end
```

Comment:
As I mentioned above, we could have a single method.

```ruby
def too_many_login_attemps?(ip: , user: )
  query = ""
  query << "IP CLAUSE" if ip
  query << "USER CLAUSE" if user
  where(query)
end
```

Query building won't work quite like that of course and `or` statements are a mess in straight AR http://stackoverflow.com/questions/18860959/rails-activerecord-where-or-clause

Given that, maybe isn't worth it.  I'm not sure how performance critical this really is, since it's not called on every request, just on login attempts.